tunnel/tunnel.go

```diff
@@ -1175,124 +1173,127 @@ func (c *Client) NewSession(target Target) (_ io.ReadWriteCloser, err error) {
 	return ioe.rwc, nil
 }
 
+// Register initializes the client register stream and determines the
+// capabilities of the tunnel server.
+func (c *Client) Register(ctx context.Context) (err error) {
+
+	c.cmu.Lock()
+	defer c.cmu.Unlock()
+	stream, err := c.tc.Register(ctx, c.cc.Opts...)
+	if err != nil {
+		return fmt.Errorf("start: failed to create register stream: %v", err)
+	}
+	c.rs = &regSafeStream{regStream: stream}
+	p, ok := peer.FromContext(stream.Context())
+	if !ok {
+		return errors.New("no peer from stream context")
+	}
+	c.addr = p.Addr
+
+	for target := range c.targets {
+		c.NewTarget(target)
+	}
+
+	for typ := range c.peerTypeTargets {
+		c.Subscribe(typ)
+	}
+
+	return nil
+}
+
 // Run initializes the client register stream and determines the capabilities
 // of the tunnel server. Once done, it starts the stream handler responsible
 // for determining what to do with received requests.
 func (c *Client) Run(ctx context.Context) error {
+	if err := c.Register(ctx); err != nil {
+		return err
+	}
+	return c.Start(ctx)
+}
+
+// Start handles received register stream requests.
+func (c *Client) Start(ctx context.Context) error {
+
 	select {
```

Comment:
You could reduce the indentation of most of the function by moving the logic from inside the select to below it.

```
select {
case c.block <- struct{}{}:
    defer func() {
        <-c.block
    }
default:
    return errors.New("client is already running")
}
// The rest of Start...
```

Reply:
Done.